### PR TITLE
feat: Support transient identities and traits

### PR DESCRIPTION
**Pull request by novakzaballa**
Closes https://github.com/Flagsmith/flagsmith-nodejs-client/issues/156

### CODE REVIEW

File: sdk/utils.ts [hunk 1]
```diff
@@ -1,12 +1,38 @@
...
 if (typeof fetch.default !== 'undefined') fetch = fetch.default;
 
-export function generateIdentitiesData(identifier: string, traits: { [key: string]: any }) {
-    const traitsGenerated = Object.entries(traits).map(trait => ({
-        trait_key: trait[0],
-        trait_value: trait[1]
-    }));
+type Trait = {
+    [key: string]: ITraitConfig | any
+}
```
Comment (khvn26) [inline sdk/utils.ts:8]: This type represents the mapping containing the identity traits, so the name should be plural, I believe.

Besides that, while it makes an illustrative point, a union with `any` means `any`, so there won't be any difference until the type guard is used.

```suggestion
type Traits = {[key: string]: any}
```

File: sdk/utils.ts [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
+
+export function isTraitConfig(trait: Trait): trait is Trait {
+    return !!trait && typeof trait == 'object' && trait.value !== undefined;
+}
```
Comment (khvn26) [inline sdk/utils.ts:12]: There's a number of problems with this type guard.

1. Judging from its name, its signature should be different:
```suggestion
export function isTraitConfig(traitValue: any): traitValue is TraitConfig {
    return !!traitValue && typeof traitValue == 'object' && traitValue.value !== undefined;
}
```

2. Judging from its usage it should accept a key-value pair, but its logic only checks the value.

File: sdk/index.ts [hunk 3]
```diff
@@ -207,11 +207,11 @@ export class Flagsmith {
...
      * @param  {string} identifier a unique identifier for the identity in the current
             environment, e.g. email address, username, uuid
-     * @param  {{[key:string]:any}} traits? a dictionary of traits to add / update on the identity in
-            Flagsmith, e.g. {"num_orders": 10}
+     * @param  {{[key:string]:any | ITraitConfig}} traits? a dictionary of traits to add / update on the identity in
+            Flagsmith, e.g. {"num_orders": 10} or {age: {value: 30, transient: true}}
      * @returns Flags object holding all the flags for the given identity.
      */
-    async getIdentityFlags(identifier: string, traits?: { [key: string]: any }): Promise<Flags> {
+    async getIdentityFlags(identifier: string, traits?: { [key: string]: any | ITraitConfig }, transient: boolean = false): Promise<Flags> {
```
Comment (khvn26) [inline sdk/index.ts:214]: Using the `any` type for trait values is problematic as a union with `any` collapses to just `any`.

Consider adding a `type IFlagsmithValue<T = string | number | boolean | null> = T` analogous to the client-side JS SDK.

File: sdk/utils.ts [hunk 4]
```diff
@@ -1,12 +1,38 @@
...
+
+export function generateIdentitiesData(identifier: string, traits: Trait, transient: boolean) {
+    const traitsGenerated = Object.entries(traits).map((trait) => {
+        if(isTraitConfig(trait)) {
+            return {
+                trait_key: trait[0],
```
Comment (khvn26) [inline sdk/utils.ts:18]: nit: Consider destructuring the map entry for better code readability.

File: tests/sdk/flagsmith-identity-flags.test.ts [hunk 5]
```diff
@@ -155,3 +155,31 @@ test('test_get_identity_flags_multivariate_value_with_local_evaluation_enabled',
...
+  fetch.mockReturnValue(Promise.resolve(new Response(identityWithTransientTraitsJSON())));
+  const identifier = 'transient_trait_identifier';
+  const traits = { some_trait: 'some_value', another_trait: {value: 'another_value', transient: true} };
+  const flg = flagsmith();
+
+  const identityFlags = (await flg.getIdentityFlags(identifier, traits)).allFlags();
+  expect(identityFlags[0].enabled).toBe(true);
+  expect(identityFlags[0].value).toBe('some-identity-with-transient-trait-value');
+  expect(identityFlags[0].featureName).toBe('some_feature');
+});
```
Comment (khvn26) [inline tests/sdk/flagsmith-identity-flags.test.ts:185]: These tests, essentially, do nothing. They mock responses and validate the mocks.

I expect them to validate the request bodies instead. In fact, if they did it, the problems I highlighted above would be caught by them.

File: flagsmith-engine/identities/models.ts [hunk 6]
```diff
@@ -29,6 +31,7 @@ export class IdentityModel {
         this.environmentApiKey = environmentApiKey;
         this.identifier = identifier;
         this.djangoID = djangoID;
+        this.transient = transient;
```
Comment (khvn26) [inline flagsmith-engine/identities/models.ts:34]: These changes are not required. These models are used exclusively for local evaluation, for which we don't need transiency info.

File: flagsmith-engine/identities/traits/models.ts [hunk 7]
```diff
@@ -1,9 +1,10 @@
 export class TraitModel {
     traitKey: string;
     traitValue: any;
-
-    constructor(key: string, value: any) {
+    transient?: boolean;
+    constructor(key: string, value: any, transient: boolean = false) {
         this.traitKey = key;
         this.traitValue = value;
+        this.transient = transient;
```
Comment (khvn26) [inline flagsmith-engine/identities/traits/models.ts:8]: Same as above.

**Review by khvn26 — CHANGES_REQUESTED** [review]
We need a number of problems fixed before I approve.

File: tests/sdk/utils.ts [hunk 8]
```diff
@@ -57,3 +57,11 @@ export function flagsJSON() {
 export function identitiesJSON() {
     return readFileSync(DATA_DIR + 'identities.json', 'utf-8');
 }
+
+export function traitIdentityJSON() {
```
[the remark this answers is not in the document]
Comment (novakzaballa) [reply tests/sdk/utils.ts]: Corrected

File: sdk/utils.ts (hunk 1 above)
Comment (novakzaballa) [reply sdk/utils.ts]: Corrected

File: sdk/utils.ts (hunk 4 above)
Comment (novakzaballa) [reply sdk/utils.ts]: Done

File: flagsmith-engine/identities/traits/models.ts (hunk 7 above)
Comment (novakzaballa) [reply flagsmith-engine/identities/traits/models.ts]: Removed

File: flagsmith-engine/identities/models.ts (hunk 6 above)
Comment (novakzaballa) [reply flagsmith-engine/identities/models.ts]: Removed

File: tests/sdk/flagsmith-identity-flags.test.ts (hunk 5 above)
Comment (novakzaballa) [reply tests/sdk/flagsmith-identity-flags.test.ts]: Corrected

File: sdk/utils.ts (hunk 2 above)
Comment (novakzaballa) [reply sdk/utils.ts]: 1) Renamed
2) Types corrected

File: flagsmith-engine/identities/traits/models.ts [hunk 9]
```diff
@@ -1,8 +1,7 @@
 export class TraitModel {
     traitKey: string;
     traitValue: any;
-
-    constructor(key: string, value: any) {
+    constructor(key: string, value: any, transient: boolean = false) {
```
Comment (khvn26) [inline flagsmith-engine/identities/traits/models.ts:4]:
```suggestion
    constructor(key: string, value: any) {
```

File: sdk/utils.ts [hunk 10]
```diff
@@ -1,12 +1,38 @@
...
+): traitValue is ITraitConfig {
+    return !!traitValue && typeof traitValue == 'object' && traitValue.value !== undefined;
+}
+
+export function generateIdentitiesData(identifier: string, traits: Traits, transient: boolean) {
+    const traitsGenerated = Object.entries(traits).map(([key, value]) => {
+        if (isTraitConfig(value)) {
+            return {
+                trait_key: key,
+                trait_value: typeof value == 'object' ? value?.value : value,
```
Comment (khvn26) [inline sdk/utils.ts:19]: Isn't the `typeof` check redundant after the type guard has been called?

```suggestion
                trait_value: value.value,
```

File: sdk/utils.ts [hunk 11]
```diff
@@ -1,12 +1,38 @@
...
+                transient: true
```
Comment (khvn26) [inline sdk/utils.ts:20]: This code is incorrect. 

```suggestion
                transient: value?.transient
```

File: tests/sdk/flagsmith-identity-flags.test.ts [hunk 12]
```diff
@@ -155,3 +155,63 @@ test('test_get_identity_flags_multivariate_value_with_local_evaluation_enabled',
...
+    {
+      trait_key:Object.keys(traits)[0],
+      trait_value:traits.some_trait
+    },
+    {
+      trait_key:Object.keys(traits)[1],
+      trait_value:traits.another_trait.value,
+      transient: true
+    }
+  ]
```
Comment (khvn26) [inline tests/sdk/flagsmith-identity-flags.test.ts:201]: Let's add an explicitly `transient: false` trait:

```suggestion
  const traits = { some_trait: 'some_value', another_trait: {value: 'another_value', transient: true, explicitly_non_transient_trait: {value: 'non_transient_value', transient: false} };
  const traitsInRequest = [
    {
      trait_key:Object.keys(traits)[0],
      trait_value:traits.some_trait,
    },
    {
      trait_key:Object.keys(traits)[1],
      trait_value:traits.another_trait.value,
      transient: true,
    },
    {
      trait_key:Object.keys(traits)[2],
      trait_value:traits.explicitly_non_transient_trait.value,
      transient: false,
    },
  ]
```

**Review by khvn26 — CHANGES_REQUESTED** [review]
Found one error, proposed one test improvement to guard against said error, and two less important suggestions.

File: sdk/utils.ts (hunk 10 above)
Comment (novakzaballa) [reply sdk/utils.ts]: Corrected

File: sdk/utils.ts (hunk 11 above)
Comment (novakzaballa) [reply sdk/utils.ts]: solved.

File: tests/sdk/flagsmith-identity-flags.test.ts (hunk 12 above)
Comment (novakzaballa) [reply tests/sdk/flagsmith-identity-flags.test.ts]: Done.

File: flagsmith-engine/identities/traits/models.ts (hunk 9 above)
Comment (novakzaballa) [reply flagsmith-engine/identities/traits/models.ts]: Removed

**Review by khvn26 — APPROVED** [review]

**Pull request merged**